### ISSUE TITLE
Dynamic libraries not linked in with `stack ghci`

### ISSUE THREAD

**Issue by mtolly** [commercialhaskell/stack#467]
Here's an example with SDL2, but this applies to any C library specified via either `extra-libraries` or (on Mac) `frameworks`.

**Steps to reproduce:**

`new-template.cabal`:

```
name:                new-template
version:             0.1.0.0
build-type:          Simple
cabal-version:       >= 1.10

library
  hs-source-dirs:      src
  exposed-modules:     Lib
  build-depends:       base >= 4.7 && < 5
  default-language:    Haskell2010
  extra-libraries:     sdl2
```

`src/Lib.hs`:

```hs
module Lib where

foreign import ccall unsafe "SDL_GetRevisionNumber"
  getRevisionNumber :: IO Int
```

`stack.yaml`:

```yaml
flags: {}
packages:
- '.'
extra-deps: []
resolver: nightly-2015-06-28
```

**Expected:**

Running `cabal repl`:

```
$ cabal repl
Preprocessing library new-template-0.1.0.0...
GHCi, version 7.8.3: http://www.haskell.org/ghc/  :? for help
Loading package ghc-prim ... linking ... done.
Loading package integer-gmp ... linking ... done.
Loading package base ... linking ... done.
Loading object (dynamic) sdl2 ... done
final link ... done
[1 of 1] Compiling Lib              ( src/Lib.hs, interpreted )
Ok, modules loaded: Lib.
*Lib λ getRevisionNumber
8628
```

**Actual:**

Running `stack ghci`:

```
$ stack ghci --verbose
Version 0.1.1.0
2015-07-01 01:45:08.728724: [debug] Checking for project config at: /Users/mtolly/git/stackbugs/stack.yaml @(stack-0.1.1.0:Stack.Config src/Stack/Config.hs:489:9)
2015-07-01 01:45:08.729279: [debug] Loading project config file stack.yaml @(stack-0.1.1.0:Stack.Config src/Stack/Config.hs:512:13)
2015-07-01 01:45:08.738228: [debug] Run process: ghc --info @(stack-0.1.1.0:System.Process.Read src/System/Process/Read.hs:253:3)
2015-07-01 01:45:08.808269: [debug] Run process: ghc --numeric-version @(stack-0.1.1.0:System.Process.Read src/System/Process/Read.hs:253:3)
2015-07-01 01:45:08.839493: [debug] Run process: ghc-pkg --no-user-package-db field --simple-output Cabal id @(stack-0.1.1.0:System.Process.Read src/System/Process/Read.hs:253:3)
2015-07-01 01:45:08.867471: [debug] Run process: ghc-pkg --no-user-package-db list --global @(stack-0.1.1.0:System.Process.Read src/System/Process/Read.hs:253:3)
2015-07-01 01:45:08.902222: [info] Configuring GHCi with the following packages: new-template @(stack-0.1.1.0:Stack.Repl src/Stack/Repl.hs:69:5)
2015-07-01 01:45:08.902683: [debug] Run process: /Users/mtolly/.stack/programs/x86_64-osx/ghc-7.10.1/bin/ghc --interactive -hide-all-packages -i/Users/mtolly/git/stackbugs/ -i/Users/mtolly/git/stackbugs/src/ -i/Users/mtolly/git/stackbugs/.stack-work/dist/x86_64-osx/Cabal-1.22.2.0/build/autogen/ -package=base /Users/mtolly/git/stackbugs/src/Lib.hs @(stack-0.1.1.0:Stack.Exec src/Stack/Exec.hs:44:5)
GHCi, version 7.10.1: http://www.haskell.org/ghc/  :? for help
[1 of 1] Compiling Lib              ( /Users/mtolly/git/stackbugs/src/Lib.hs, interpreted )

ByteCodeLink: can't find label
During interactive linking, GHCi couldn't find the following symbol:
  SDL_GetRevisionNumber
This may be due to you not asking GHCi to load extra object files,
archives or DLLs needed by your current session.  Restart GHCi, specifying
the missing library using the -L/path/to/object/dir and -lmissinglibname
flags, or simply by naming the relevant files on the GHCi command line.
Alternatively, this link failure might indicate a bug in GHCi.
If you suspect the latter, please send a bug report to:
  glasgow-haskell-bugs@haskell.org

 λ
```

Here's how to fix it manually:

```
$ stack ghci --ghc-options -lsdl2
Configuring GHCi with the following packages: new-template
GHCi, version 7.10.1: http://www.haskell.org/ghc/  :? for help
[1 of 1] Compiling Lib              ( /Users/mtolly/git/stackbugs/src/Lib.hs, interpreted )
Ok, modules loaded: Lib.
*Lib λ getRevisionNumber
8628
```

**Comment by chrisdone** [comment]
Want to take a bash at fixing this? It'll be in here: https://github.com/commercialhaskell/stack/blob/master/src/Stack/Package.hs#L295..L321

**Comment by chrisdone** [comment]
Merged as 542f027, can you confirm and close this issue? Sorry about the delay, I moved onto a high prio project.

**Comment by exarkun** [comment]
This appears to be broken on `1.9.3 x86_64 hpack-0.31.2`.  I can reproduce what seems to be the same issue against https://github.com/PrivateStorageio/PaymentServer/tree/25ffdca08dd89e90d6cecd2a13147a92dd820c94 on NixOS 19.03:

```
*Main PaymentServer.Issuer PaymentServer.Main PaymentServer.Persistence PaymentServer.Processors.Stripe PaymentServer.Redemption PaymentServer.Ristretto PaymentServer.Server> :l PaymentServer.Ristretto 
*Main PaymentServer.Issuer PaymentServer.Main PaymentServer.Persistence PaymentServer.Processors.Stripe PaymentServer.Redemption PaymentServer.Ristretto PaymentServer.Server> :l PaymentServer.Ristretto 
ghc: ^^ Could not load 'blinded_token_decode_base64', dependency unresolved. See top entry above.


ByteCodeLink: can't find label
During interactive linking, GHCi couldn't find the following symbol:
  blinded_token_decode_base64
This may be due to you not asking GHCi to load extra object files,
archives or DLLs needed by your current session.  Restart GHCi, specifying
the missing library using the -L/path/to/object/dir and -lmissinglibname
flags, or simply by naming the relevant files on the GHCi command line.
Alternatively, this link failure might indicate a bug in GHCi.
If you suspect the latter, please send a bug report to:
  glasgow-haskell-bugs@haskell.org
```

However, using `stack run` causes the dynamic library to be linked in fine.
